onnxruntime/core/providers/webnn/builders/impl/gemm_op_builder.cc

```diff
@@ -162,6 +157,29 @@ bool GemmOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers,
     }
   }
 
+  if (op_type == "MatMul") {
+    if (a_shape.size() < 2 || b_shape.size() < 2) {
+      LOGS(logger, VERBOSE) << "Inputs of MatMul must be at least 2D";
+      return false;
+    }
+
+    // WebNN CPU backend has two more constraints.
+    // https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ml/webnn/ml_graph_xnnpack.cc;l=1177
+    if (device_type == WebnnDeviceType::CPU) {
```

Comment:
The Chromium WebNN XNNPack backend is supposed to handle this by setting the rank of the input tensors to be consistent before XNNPack even sees it (similar to the DML Backend which has this exact same constraint). It shouldn't be handled by higher level frameworks like TF Lite and ORT. Is there a todo to fix this properly on Chromium?

Reply:
There should be a TODO in native XNNPack to support these constraints. Cc/ @huningxin.